File: 283-move-zeroes/move-zeroes.py
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # use two pointers, both  at beginning and then compare current with the next
        # if theres a zero then swap current index with previous index





        # Start search_zero at 1
        search_zero = 0

        for i in range(len(nums)):
            if nums[i] != 0:
                # Swap the current element with the previous non-zero element
                nums[i], nums[search_zero] = nums[search_zero], nums[i]
                search_zero += 1


        

File: 283-move-zeroes/move-zeroes.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var moveZeroes = function(nums) {
  let left = 0;   // next spot to place a non-zero
  let right = 0;  // scans through the array

  while (right < nums.length) {
    if (nums[right] !== 0) {
      // swap only when needed
      if (left !== right) {
        [nums[left], nums[right]] = [nums[right], nums[left]];
      }
      left++; // we placed a non-zero; next slot moves forward
    }
    right++;  // always scan forward
  }


    
};